kernel/src/syscall.rs

```diff
@@ -394,12 +395,40 @@ pub enum SyscallReturnVariant {
     FailureU32 = 1,
     FailureU32U32 = 2,
     FailureU64 = 3,
+    FailurePtrUsize = 4,
+    FailurePtrPtr = 5,
     Success = 128,
     SuccessU32 = 129,
     SuccessU32U32 = 130,
     SuccessU64 = 131,
     SuccessU32U32U32 = 132,
     SuccessU32U64 = 133,
+    SuccessUsize = 134,
+    SuccessPtr = 135,
+    SuccessPtrUsize = 136,
+    SuccessPtrPtr = 137,
+}
```

Comment:
This can be split into a trd104 enum and a new enum to clarify what is part of the syscall ABI.

Reply:
I certainly can define both enums, but I am not sure how you would like them used as we would, at each point in code, need to actually pick one to use.

The intent here is to allow capsules / the kernel access to new variants on platforms they exist and so I presume they would all use the new one. The only time we would want the trd104 one was during lowering, but the lowering function lowers to u32 (or usize), so it would not be an Enum there either. I can create the enum, just for reference, but I don't think it would find use.

Reply:
This PR is tricky as it tries to adjust this infrastructure for a future PR that would make use of these new variants over the system call interface (in a different `encode_syscall_return` function), but this PR is changing these existing types, despite these new variants actually being supposed to be propagated to userspace.

If creating a new type doesn't make sense, can we just exclude them from this PR, or comment out?

At this point I really feel like it's so much easier to just define two entirely separate `encode_syscall_return` functions that will each use their respective `SyscallReturnVariant` enums), and not change this type in this PR.

Reply:
> but this PR is changing these existing types, despite these new variants actually being supposed to be propagated to userspace.
> can we just exclude them from this PR, or comment out?

They are being used (and so commenting them out would fail to compile).
Although this PR does not propagate them to userspace on 32-bit platforms, it does on 64-bit platforms, and so sets the stage for one to be introduced.

I am not sure how we can both set up ground kernel ground work for 64-bit RISCV, but not yet include anything that it would motivate.

Reply:
Ah, didn't spot this comment. I gave an outline of why I prefer the single implementation in a comment below.   I do see the value of having type system guarantees. I think we can maybe have both that guarantee and less code duplication?

On some level, I will point out, we are not actually giving the guarantee of getting return codes correct for trd104 systems using the extra syscall type / syscall variant enum. We can still mis-define those types, return the wrong trd104 code, and given we lower to enum to usize in the function, we could still accidentally return _any_ usize. At some level, we are just relying on logical correctness because our lowering is to usize, and just relying on the variant map function to definitely only return old codes is not any worse in my mind. It is a pretty trivial code review to check all the RHS of the match contain only old codes.

We can also just get rid of the pointer casts by assigning to a temporary and using "as", if that was the object of the trd104 wrapper. No unsafe required.

Reply:
https://github.com/tock/tock/pull/4174#discussion_r1841454858

Responded to the other thread. Now that this is split out of this PR, let's defer (continuing) this discussion until the follow-up PR is up and we can see how this is used.